File: core-presentation/src/main/java/home/oleg/placesnearme/core_presentation/base/BaseViewModel.java
package home.oleg.placesnearme.core_presentation.base;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.smedialink.common.function.Action;

import java.util.LinkedList;
import java.util.Queue;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

/**
 * Created by Oleg Sheliakin on 18.09.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */
public abstract class BaseViewModel<VIEW extends View> extends ViewModel {

    private MutableLiveData<Action<VIEW>> observer;

    private final CompositeDisposable compositeDisposable = new CompositeDisposable();

    private final Queue<Action<VIEW>> queue = new LinkedList<>();

    public MutableLiveData<Action<VIEW>> getObserver() {
        if(observer == null) {
            observer = new MutableLiveData<>();
            onObserverCreated();
        }
        return observer;
    }

    protected void onObserverCreated() {

    }

    protected boolean addToDisposables(Disposable disposable) {
        return compositeDisposable.add(disposable);
    }

    protected void setState(Action<VIEW> action) {
        observer.setValue(action);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.clear();
    }
}

File: common/src/main/java/com/smedialink/common/base/BaseViewModel.kt
package com.smedialink.common.base

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.smedialink.common.ext.reduceState
import androidx.lifecycle.ViewModel
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

/**
 * Created by Oleg Sheliakin on 06.11.2018.
 * Contact me by email - olegsheliakin@gmail.com
 */

abstract class BaseViewModel<STATE : Any> : ViewModel() {

    protected val disposables: CompositeDisposable by lazy {
        return@lazy CompositeDisposable()
    }

    protected val stateInternal: MutableLiveData<STATE> by lazy {
        return@lazy MutableLiveData<STATE>()
    }

    val state: LiveData<STATE> = stateInternal

    override fun onCleared() {
        super.onCleared()
        disposables.clear()
    }

    protected fun Disposable.autoDispose() : Disposable {
        disposables.add(this)
        return this
    }

    protected fun reduce(operation: (STATE) -> STATE) {
        stateInternal.reduceState {
            operation(it)
        }
    }
}
